### PR TITLE
Support multiple metrics in model training

### PR DESCRIPTION
**Pull request by futurely**
This closes https://github.com/dmlc/mxnet/issues/751, https://github.com/dmlc/mxnet/issues/1362 and is related to https://github.com/dmlc/mxnet/issues/269.

### CODE REVIEW

File: python/mxnet/callback.py [hunk 1]
```diff
@@ -81,9 +81,16 @@ def __call__(self, param):
             if count % self.frequent == 0:
                 speed = self.frequent * self.batch_size / (time.time() - self.tic)
                 if param.eval_metric is not None:
+                    logging.info("Epoch[%d] Batch [%d]\tSpeed: %.2f samples/sec",
+                                 param.epoch, count, speed)
                     name, value = param.eval_metric.get()
-                    logging.info("Epoch[%d] Batch [%d]\tSpeed: %.2f samples/sec\tTrain-%s=%f",
-                                 param.epoch, count, speed, name, value)
+                    if isinstance(name, list) and isinstance(value, list) and \
```
Comment (piiswrong) [inline python/mxnet/callback.py]: also add this to log_train_metric?

File: python/mxnet/model.py [hunk 2]
```diff
@@ -155,7 +155,7 @@ def _train_multi_device(symbol, ctx, arg_names, param_names, aux_names,
     eval_data : DataIter
         Validation data iterator.
     eval_metric : EvalMetric
-        A evaluation function.
+        One or more evaluation functions.
```
Comment (piiswrong) [inline python/mxnet/model.py]: "a evaluation function or a list of evaluation functions"

File: python/mxnet/metric.py [hunk 3]
```diff
@@ -58,6 +57,45 @@ def get(self):
         else:
             return (self.name, self.sum_metric / self.num_inst)
 
+class CompositeEvalMetric(EvalMetric):
+    """Manage multiple evaluation metrics."""
+
+    def __init__(self):
```
Comment (piiswrong) [inline python/mxnet/metric.py]: support initializing with a list of metric for convinience.

File: python/mxnet/model.py [hunk 4]
```diff
@@ -303,6 +303,9 @@ def _train_multi_device(symbol, ctx, arg_names, param_names, aux_names,
                     else:
                         eval_batch_end_callback(batch_end_params)
             name, value = eval_metric.get()
+            if isinstance(name, list) and isinstance(value, list) and len(name) == len(value):
+                for n, v in zip(name, value):
+                    logger.info('Epoch[%d] Validation-%s=%f', epoch, n, v)
             logger.info('Epoch[%d] Validation-%s=%f', epoch, name, value)
```
Comment (piiswrong) [inline python/mxnet/model.py]: else?

**Comment by piiswrong** [comment]
THanks for the contribution. I added a few comments. Did you verify that this works?

**Comment by futurely** [comment]
Yes, it works.

File: python/mxnet/metric.py [hunk 5]
```diff
@@ -65,20 +106,27 @@ def get(self):
...
+        except KeyError:
+            self.top_k = 1
+        self.name += '-Top-%d' % self.top_k
 
     def update(self, labels, preds):
         check_label_shapes(labels, preds)
 
         for i in range(len(labels)):
-            pred_label = ndarray.argmax_channel(preds[i]).asnumpy().astype('int32')
+            pred_label = numpy.argsort(preds[i].asnumpy().astype('float32'), axis = 1)
```
Comment (piiswrong) [inline python/mxnet/metric.py]: ndarray.argmax_channel allows computing prediction on GPU, this is critical for performance when output is large, e.g. for segmentation.
Please keep it unless top_k > 1.

**Comment by futurely** [comment]
Done.

File: python/mxnet/metric.py [hunk 6]
```diff
@@ -65,20 +116,36 @@ def get(self):
...
+                pred_label = ndarray.argmax_channel(preds[i]).asnumpy().astype('int32')
             label = labels[i].asnumpy().astype('int32')
 
             check_label_shapes(label, pred_label)
-
-            self.sum_metric += (pred_label.flat == label.flat).sum()
-            self.num_inst += len(pred_label.flat)
+            num_samples = pred_label.shape[0]
+            num_dims = len(pred_label.shape)
+            if num_dims == 1:
```
Comment (piiswrong) [inline python/mxnet/metric.py]: check top_k here instead of num_dims.
pred_label can have multiple dimensions for multi label prediction tasks.

**Comment by piiswrong** [comment]
Looks good. Somehow scalatest is blocking. We can merge after it's resolved.

Please merge in the master

**Comment by tqchen** [comment]
This is merged. Thanks!

**Pull request merged**